### PR TITLE
fix: Explicitly set the rel=canonical of the OpenTelemetry blog post

### PR DESCRIPTION
**Pull request by nfrankel**
Duplicate content is frowned upon by Google. This blog post has been already published on the OpenTelemetry web site. To avoid being flagged as duplicate and incur a SEO penalty, we need to reference the "original".

Alternatively, we could ask the OpenTelemetry team to point to our post. But it will probably take more time and it's not worth it IMHO.

### CODE REVIEW

**Comment by GeekEmily** [comment]
Looks good to me

**Review by yzeng25 — APPROVED** [review]

**Comment by SkyeYoung** [comment]
I don't think it's a good idea. Canonical Tags specify the URL that should be shown in search results. It should be used to specify the version of a page in our website, not other websites.

**Comment by nfrankel** [comment]
@SkyeYoung It makes sense. Are you willing to ask (and follow up) the OpenTelemetry team to update their page?

**Comment by yzeng25** [comment]
Technically they are two articles with different titles and contents. Adding it wouldn't hurt, but I think it is also fine to leave it as it is.

https://opentelemetry.io/blog/2022/apisix/
https://apisix.apache.org/blog/2022/02/28/apisix-integration-opentelemetry-plugin

**Comment by SkyeYoung** [comment]
@yzeng25 No. Canonical Tags mean which URL should be indexed. If merged this, https://apisix.apache.org/blog/2022/02/28/apisix-integration-opentelemetry-plugin may never be indexed.

**Comment by nfrankel** [comment]
@yzeng25 Technically, I'm afraid that Google might disagree with you 😅

> Duplicate content generally refers to substantive blocks of content within or across domains that either completely match other content in the same language or **are appreciably similar**
>
>-- [Avoid creating duplicate content](https://developers.google.com/search/docs/advanced/guidelines/duplicate-content)

(Emphasis is mine)

**Comment by SkyeYoung** [comment]
@nfrankel @yzeng25 we can read the [HTML Standard](https://html.spec.whatwg.org/multipage/links.html#link-type-canonical)

**Comment by nfrankel** [comment]
>Canonical Tags mean which URL should be indexed

Can you please provide references to this statement @SkyeYoung

**Comment by nfrankel** [comment]
>we can read the [HTML Standard](https://html.spec.whatwg.org/multipage/links.html#link-type-canonical)

Unfortunately, it has nothing to do with HTML standard but everything to do with how Google indexes documents

**Comment by SkyeYoung** [comment]
> > Canonical Tags mean which URL should be indexed
> 
> Can you please provide references to this statement @SkyeYoung

@nfrankel Pls read https://datatracker.ietf.org/doc/html/rfc6596.


> Unfortunately, it has nothing to do with HTML standard but everything to do with how Google indexes documents

We may need to know the members of WHATWG. :D

**Comment by nfrankel** [comment]
>Authors who declare the canonical link relation ought to anticipate that applications such as search engines **can**:
>
>* Index content only from the target IRI (i.e., content from the context IRIs will be likely disregarded as duplicative).

(Emphasis mine).

@SkyeYoung If I understand correctly, your point is that if we set the `rel=canonical`, our page **may** not be indexed.

My point is that I don't care, because it **will** be flagged as duplicate by Google, thus our domain **will** lose PageRank or whatever secret metric Google uses nowadays and our page **may** not be indexed anyway.

I stand by my position: we need to set the `rel=canonical`.

**Comment by yzeng25** [comment]
So is it better to add Canonical on OpenTelemetry's side? Or adding Canonical is a bad idea for both sides? @nfrankel @SkyeYoung

**Comment by nfrankel** [comment]
The best would definitely be to add the canonical on their side and point to our blog (as I mentioned in the PR description).

**Comment by SkyeYoung** [comment]
> My point is that I don't care, because it **will** be flagged as duplicate by Google, thus our domain **will** lose PageRank or whatever secret metric Google uses nowadays and our page **may** not be indexed anyway.

**Duplicate**  just say on the **same** website.

**Comment by yzeng25** [comment]
> The best would definitely be to add the canonical on their side and point to our blog (as I mentioned in the PR description).

Say if I had a pull request opening in OpenTelemetry's side, and I'll add something like this to it.

```
`<head>
    <link rel="canonical" href="https://apisix.apache.org/blog/2022/02/28/apisix-integration-opentelemetry-plugin//" />
</head>`
```

And that is it, right?

**Comment by nfrankel** [comment]
>Duplicate just say on the same website

No!!!

Please read my [above comment](https://github.com/apache/apisix-website/pull/995#issuecomment-1084277029). To make it easier, I'll again paste and emphasize the important bits:

>Duplicate content generally refers to substantive blocks of content within or **across domains** that either completely match other content in the same language or are appreciably similar
>
>-- [Duplicate content](https://developers.google.com/search/docs/advanced/guidelines/duplicate-content)

**Comment by nfrankel** [comment]
@yzeng25 I don't know if they are using Docusaurus too, but if they are, yes, something like that could work. But in the meanwhile, Google would flag us as duplicate. So it depends a lot on how fast they can merge the PR.

**Comment by yzeng25** [comment]
@nfrankel adding the above code block to this pull request should solve all problems, right?

https://github.com/open-telemetry/opentelemetry.io/pull/1254

**Comment by nfrankel** [comment]
@yzeng25 Nope

>This is the source repo for the website and project documentation. The site is using **Hugo** and hosted on Netlify.
>
>-- https://github.com/open-telemetry/opentelemetry.io

We need to understand:

1. If we can override the canonical link with Hugo
2. How to do it

**Comment by nfrankel** [comment]
I had a [quick look](https://gohugo.io/content-management/urls/) and it's a bit involved. It seems you cannot just add a snippet: you'd need to actually to [change the template](https://cconcannon.medium.com/adding-an-external-canonical-url-to-a-hugo-template-3b6e18814649).

**Comment by yzeng25** [comment]
> @yzeng25 Nope
> 
> > This is the source repo for the website and project documentation. The site is using **Hugo** and hosted on Netlify.
> > -- https://github.com/open-telemetry/opentelemetry.io
> 
> We need to understand:
> 
> 1. If we can override the canonical link with Hugo
> 2. How to do it

There you go, I think it is totally feasible.

image

**Comment by nfrankel** [comment]
Amazing @yzeng25! Then by all means, please open a PR. But let's watch it, and if it's not merged within a reasonable time frame, we shall get back to this one.

**Comment by SkyeYoung** [comment]
> Please read my [above comment](https://github.com/apache/apisix-website/pull/995#issuecomment-1084277029). To make it easier, I'll again paste and emphasize the important bits:
> 
> > Duplicate content generally refers to substantive blocks of content within or **across domains** that either completely match other content in the same language or are appreciably similar
> > -- [Duplicate content](https://developers.google.com/search/docs/advanced/guidelines/duplicate-content)

Sorry, I didn't see that. Interesting. They don't seem to like following their own standards.

But, I found:

> Duplicate content on a site is not grounds for action on that site unless it appears that the intent of the duplicate content is to be deceptive and manipulate search engine results. If your site suffers from duplicate content issues, and you don't follow the advice listed in this document, we do a good job of choosing a version of the content to show in our search results.
>
> -- [Duplicate content](https://developers.google.com/search/docs/advanced/guidelines/duplicate-content)

Maybe we don't have to do anything

**Comment by nfrankel** [comment]
>They don't seem to like following their own standards

They don't have "standards". They are Google. They do whatever they want.

>Maybe we don't have to do anything

Definitely no. I'm not an SEO expert, but I've dabbled a bit into it, and duplicate content is just plain bad.

**Comment by SkyeYoung** [comment]
> They don't have "standards". They are Google. They do whatever they want.

So I use DDG, but not Google. :p

**Comment by nfrankel** [comment]
DDG?

**Comment by SkyeYoung** [comment]
@nfrankel Duck Duck Go :D

**Comment by yzeng25** [comment]
> > @yzeng25 Nope
> > > This is the source repo for the website and project documentation. The site is using **Hugo** and hosted on Netlify.
> > > -- https://github.com/open-telemetry/opentelemetry.io
> > 
> > 
> > We need to understand:
> > 
> > 1. If we can override the canonical link with Hugo
> > 2. How to do it
> 
> There you go, I think it is totally feasible.
> 
> image

@nfrankel @SkyeYoung as I understand, this is the most satisfying way for all of us, right?

**Comment by nfrankel** [comment]
@yzeng25 yes, if it's merged fast

**Comment by nfrankel** [comment]
@SkyeYoung Ah yes. Unfortunately (or not), Google is ubiquitous everywhere outside of China (and perhaps Russia with Yandex)

**Comment by yzeng25** [comment]
@nfrankel @SkyeYoung I have it done [here](https://github.com/open-telemetry/opentelemetry.io/pull/1254/commits/cda39aab4149ebd5d18539e21939a776aaa5a749). Normally it takes a couple of days. I will track it.

**Comment by SkyeYoung** [comment]
Good job!
________________________________
发件人: yilinzeng ***@***.***>
发送时间: Thursday, March 31, 2022 5:58:51 PM
收件人: apache/apisix-website ***@***.***>
抄送: SkyeYoung ***@***.***>; Mention ***@***.***>
主题: Re: [apache/apisix-website] fix: Explicitly set the rel=canonical of the OpenTelemetry blog post (PR #995)


@nfrankel<https://github.com/nfrankel> @SkyeYoung<https://github.com/SkyeYoung> I have it done here<https://github.com/open-telemetry/opentelemetry.io/pull/1254/commits/cda39aab4149ebd5d18539e21939a776aaa5a749>. Normally it takes a couple of days. I will track it.

―

**Comment by nfrankel** [comment]
Thanks a lot @yzeng25!

**Comment by yzeng25** [comment]
> Thanks a lot @yzeng25!

No problem. Do you mind converting this pull request to draft for now? I want to see how it goes on the other side first. In this way, this pull request won't get merged or need re-approval.

**Comment by yzeng25** [comment]
https://github.com/open-telemetry/opentelemetry.io/pull/1254#discussion_r839675423 solves part of our concerns in this issue.

This pull request will solve the rest of our concerns. 

https://github.com/open-telemetry/opentelemetry.io/pull/1258

**Comment by yzeng25** [comment]
@nfrankel @SkyeYoung, since both https://github.com/open-telemetry/opentelemetry.io/pull/1258 and https://github.com/open-telemetry/opentelemetry.io/pull/1254 are merged now, canonical url is added on OpenTelemetry's side. See https://opentelemetry.io/blog/2022/apisix/#summary.

I think we can close this pull request.

**Comment by nfrankel** [comment]
Sorry, but I don't see any canonical link in the deployed page...

**Comment by yzeng25** [comment]
> Sorry, but I don't see any canonical link in the deployed page...

It is here.

https://raw.githubusercontent.com/open-telemetry/opentelemetry.io/main/content/en/blog/2022/apisix.md

image

image

**Comment by SkyeYoung** [comment]
@yzeng25 @nfrankel It seems to be more like a variable, and not parsed to canonical tag

**Comment by yzeng25** [comment]
Ok. Take Medium as an example, after adding canonical_url, we are expecting something like this to show up.

image

I checked all blog posts on OpenTelemetry's site, all blogs with `canonical_url` are not reflecting similar changes as above.

image

So we need to go back to the second option, add canonical_url via this pull request, right?

**Comment by SkyeYoung** [comment]
@yzeng25 Maybe they just have not intended to support Canonical Tag. 🤔

I remember that hugo did not have related FrontMatter (I used to be a user of it)

**Comment by yzeng25** [comment]
> @yzeng25 Maybe they just have not intended to support Canonical Tag. 🤔
> 
> I remember that hugo did not have related FrontMatter (I used to be a user of it)

This is hilarious 😂。I spend hours to work on it and ask their community to merge pull requests, and it is not working as expected.

Positive side: we found a bug for them.

Negative side: we need an alternative to solve canonical_url issue.

**Comment by SkyeYoung** [comment]
> Positive side: we found a bug for them.

Maybe this is the behavior they expect...

**Comment by nfrankel** [comment]
@yzeng25 Please merge this PR. We tried, we failed. Now it's time to fix.

**Comment by yzeng25** [comment]
It is a completely new case for me. As we discussed, tried and encountered failures, let's agree to disagree. 

1. I would like to get this pull request merged. Giving it a shot wouldn't hurt.
2. We should review GA stats after a month or so, to see if `canonical_url` actually works. If it works, then we should do it for each cross-posted blog that is posted on platforms other than Medium. If it does not work,  we should rollback the change in this pull request.

**Comment by SkyeYoung** [comment]
> 2\. We should review GA stats after a month or so, to see if `canonical_url` actually works. If it works, then we should do it for each cross-posted blog that is posted on platforms other than Medium. If it does not work,  we should rollback the change in this pull request.

Let's create an issue to track this.

**Comment by nfrankel** [comment]
@yzeng25 Perfect for me

**Comment by yzeng25** [comment]
> > 2. We should review GA stats after a month or so, to see if `canonical_url` actually works. If it works, then we should do it for each cross-posted blog that is posted on platforms other than Medium. If it does not work,  we should rollback the change in this pull request.
> 
> Let's create an issue to track this.

@nfrankel @SkyeYoung , just created an issue to track this: https://github.com/apache/apisix-website/issues/999. We can discuss over there.

**Review by SkyeYoung — APPROVED** [review]
LGTM

**Review by SylviaBABY — APPROVED** [review]

**Review by guitu168 — APPROVED** [review]
LGTM

**Pull request merged**
